geoips/commandline/geoips_command.py

```diff
@@ -0,0 +1,357 @@
+                self.cmd_instructions = parent.cmd_instructions
+            else:
+                # Otherwise use the default cmd_instructions which are used for normal
+                # invocation of the CLI.
+                self.cmd_instructions = cmd_instructions
+            try:
+
+                # attempt to create a sepate sub-parser for the specific sub-command
+                # class being initialized
+                # So we can separate the commands arguments in a tree-like structure
+                self.subcommand_parser = parent.subparsers.add_parser(
+                    self.command_name,
+                    help=self.cmd_instructions["instructions"][self.combined_name][
+                        "help_str"
+                    ],
+                    usage=self.cmd_instructions["instructions"][self.combined_name][
+                        "usage_str"
+                    ],
+                )
+            except KeyError:
+                raise KeyError(
+                    "Error, the supplied command line instructions are improperly "
+                    "formatted. You need an 'instructions' entry that contains a "
+                    f"'{self.combined_name}' key."
+                )
+        else:
+            # otherwise initialize a top-level parser for this command.
+            self.subcommand_parser = argparse.ArgumentParser()
+            self.combined_name = self.command_name
+
+        self.add_subparsers()
+
+    @property
+    @abc.abstractmethod
+    def command_name(self):
+        """Name of the command class."""
+        pass
+
+    @property
+    @abc.abstractmethod
+    def subcommand_classes(self):
+        """List of subcommand_classes related to the top level command.
+
+        For example, if the class provided was GeoipsList, subcommand_classes would
+        be the list of available subcommand_classes that "geoips list" implements.
+        """
+        pass
+
+    def add_subparsers(self):
+        """Add subparsers for each sub-command class.
+
+        This is done so we can limit the scope of what arguments are accepted for each
+        geoips <cmd> sub-command. This is only done for the top-level command, such as
+        "list", "run", "get", etc.
+
+        For example, if this were the GeoipsList Command Sub-Class, we would create a
+        self.list_subparsers attribute, which we then add individual parsers for each
+        sub-command, as in interfaces, plugins, packages, scripts, etc.
+        """
+        if len(self.subcommand_classes):
+            self.subparsers = self.subcommand_parser.add_subparsers(
+                help=f"{self.command_name} instructions."
+            )
+            for subcmd_cls in self.subcommand_classes:
+                subcmd_cls(parent=self)
+
+    @property
+    def plugin_package_names(self):
+        """List of names of all installed Geoips Plugin Packages."""
+        return plugin_packages.entrypoints
```

Comment:
Maybe this should be sorted by package name. The usage output from calling `geoips get package -h` is unordered and difficult to read:

```
> geoips get package -h
usage: To use, type `geoips get package <package_name>`, where <package_name> is a
valid GeoIPS Package.

positional arguments:
  {my_package,geoips_clavrx,my_fusion_package,data_fusion,recenter_tc,geoips,geoips_plugin_example}
                        GeoIPS Package to retrieve.

options:
  -h, --help            show this help message and exit
```

Maybe create an issue to find all of the similar properties and make them ordered.

Reply:
I can either place a ``sorted()`` call around ``PluginPackages.entrypoints`` or I can create a separate issue for this @jsolbrig. IMO, this is an easy fix and can likely just be included in this PR. Up to you though.

Reply:
This issue has been fixed in the PluginPackages class. We call ``sorted(get_entry_point_group("geoips.plugin_packages"))`` which sorts the entry points by their name / value. We do this for both ``PluginPackages.entrypoints`` and ``PluginPackages.paths``.

Reply:
Resolving this issue.